astroquery/cadc/cadctap/core.py

```diff
@@ -0,0 +1,409 @@
+                           login_url=login_url, verbose=verbose)
+        else:
+            if password is None:
+                raise AttributeError("No password")
+                return
+
+    def __dologin(self, cookie_prefix=None, login_url=None, verbose=False):
+        """
+        Notes
+        -----------------
+        Decode and set the cookie with the right key
+        """
+        self._TapPlus__isLoggedIn = False
+        response = self.__execLogin(self._TapPlus__user,
+                                    self._TapPlus__pwd,
+                                    login_url,
+                                    verbose)
+        # check response
+        connHandler = self._TapPlus__getconnhandler()
+        isError = connHandler.check_launch_response_status(response,
+                                                           verbose,
+                                                           200)
+        if isError:
+            if verbose:
+                print("Login error: " + str(response.reason))
+            raise requests.exceptions.HTTPError(
+                "Login error: " + str(response.reason))
+        else:
+            # extract cookie
+            cookie = response.read()
+            c = cookie.decode()
+            cookie = '{}="{}"'.format(cookie_prefix, c)
+            if cookie is not None:
+                self._TapPlus__isLoggedIn = True
+                connHandler.set_cookie(cookie)
+
+    def logout(self, verbose=False):
+        """Performs a logout
+        Parameters
+        ----------
+        verbose : bool, optional, default 'False'
+            flag to display information about the process
+
+        Notes
+        -----------------
+        Clear the certificate as well
+        """
+        self._TapPlus__getconnhandler()._TapConn__connectionHandler. \
+            _ConnectionHandlerCadc__certificate = None
+        if self._TapPlus__getconnhandler().cookies_set():
+            self._TapPlus__getconnhandler().unset_cookie()
+        self.__certificate = None
+        self._TapPlus__isLoggedIn = False
+
+    def __execLogin(self, usr, pwd, login_url=None, verbose=False):
+        """
+        Notes
+        -----------------
+        Send to the right login url, send not using the TapConn object
+        because it redirects instead of giving you a cookie
+        """
+        args = {
+            "username": str(usr),
+            "password": str(pwd)}
+        data = urlencode(args)
+        if login_url is None:
+            raise RuntimeError("No login URL")
+        protocol, host, port, server_context, \
+            tap_context = self._Tap__parseUrl(login_url)
+        connHandler = httplib.HTTPConnection(host, 80)
+        context = '/ac/login'
+        header = {
+            "Content-type": "application/x-www-form-urlencoded",
+            "Accept": "text/plain"
+        }
+        connHandler.request("POST", context, data, header)

```

Comment:
this should use astroquery's builtin methods (`Query._request`, for example)

Reply:
if this is needed for some reason - e.g. because of something TAP is doing - we need to refactor TAP.

Reply:
See my comment above.